File: src/core/net/strategies/format-rotation.test.ts
import { describe, expect, it } from 'vitest'
import { createComboCache, interfaceKeyOf } from '../combo'
import { createNgaFetcher } from '../fetcher'
import type { HttpRequest, HttpResponse, HttpTransport } from '../transport'
import { createFormatRotationStrategy } from './format-rotation'

const utf8 = (text: string) => new TextEncoder().encode(text)

const OK = utf8('{"data":{"0":"ok"},"time":1}')
/** 被封的典型表现：返回一坨 HTML，洗不成 JSON（ADR-0002） */
const BLOCKED = utf8('<html><body>403 Forbidden</body></html>')

const HOSTS = ['https://bbs.nga.cn', 'https://ngabbs.com'] as const
const FORMATS = ['json', 'jsonLite'] as const

/** 组合的可读名，断言顺序时比生 URL 好读。 */
function comboOf(request: HttpRequest): string {
  const url = new URL(request.url)
  const lite = url.searchParams.get('lite')
  const format = lite === null ? `__output=${url.searchParams.get('__output')}` : `lite=${lite}`
  return `${format}@${url.origin}`
}

function fakeTransport(
  respond: (combo: string) => Partial<HttpResponse>,
): { transport: HttpTransport; combos: string[] } {
  const combos: string[] = []
  const transport: HttpTransport = (request) => {
    const combo = comboOf(request)
    combos.push(combo)
    const response = respond(combo)
    return Promise.resolve({
      status: response.status ?? 200,
      // 这个文件里的响应体都是 UTF-8 字面量（GBK 解码另有 encoding/ 的单测钉着）
      contentType: response.contentType ?? 'text/javascript; charset=UTF-8',
      body: response.body ?? OK,
    })
  }
  return { transport, combos }
}

/** 只有 `only` 这个组合是通的，其余一律返回封禁页。 */
function onlyWorking(only: string) {
  return fakeTransport((combo) =>
    combo === only ? { body: OK } : { status: 403, body: BLOCKED, contentType: 'text/html' },
  )
}

function rotatingFetcher(transport: HttpTransport, cache = createComboCache()) {
  return createNgaFetcher({
    transport,
    comboCache: cache,
    host: HOSTS[0],
    strategies: [
      createFormatRotationStrategy({ formats: FORMATS, hosts: HOSTS, maxAttempts: 10 }),
    ],
  })
}

describe('createFormatRotationStrategy · 格式参数 × 域名的组合枚举', () => {
  it('封禁响应触发按序降级,直到某个组合通了', async () => {
    const { transport, combos } = onlyWorking('lite=js@https://ngabbs.com')
    const fetchNga = rotatingFetcher(transport)

    const result = await fetchNga({ path: 'thread.php', query: { fid: 650 } })

    expect(result.via).toBe('format-rotation')
    expect(combos).toEqual([
      '__output=8@https://bbs.nga.cn',
      'lite=js@https://bbs.nga.cn',
      '__output=8@https://ngabbs.com',
      'lite=js@https://ngabbs.com',
    ])
  })

  it('成功的组合按接口 key 记进缓存,下一次直接命中不再从头试', async () => {
    const cache = createComboCache()
    const { transport, combos } = onlyWorking('lite=js@https://ngabbs.com')
    const fetchNga = rotatingFetcher(transport, cache)
    const request = { path: 'thread.php', query: { fid: 650 } }

    await fetchNga(request)
    expect(cache.get(interfaceKeyOf(request))).toEqual({
      format: 'jsonLite',
      host: 'https://ngabbs.com',
    })

    combos.length = 0
    await fetchNga(request)

    expect(combos).toEqual(['lite=js@https://ngabbs.com'])
  })

  it('缓存是按接口分的:另一个接口还得自己从头试', async () => {
    const cache = createComboCache()
    const { transport, combos } = onlyWorking('lite=js@https://bbs.nga.cn')
    const fetchNga = rotatingFetcher(transport, cache)

    await fetchNga({ path: 'nuke.php', query: { __lib: 'noti', __act: 'get_all' } })
    combos.length = 0
    await fetchNga({ path: 'nuke.php', query: { __lib: 'ucp', __act: 'get' } })

    expect(combos).toEqual(['__output=8@https://bbs.nga.cn', 'lite=js@https://bbs.nga.cn'])
  })

  it('全组合都被封时,缓存里那个也清掉,免得下次还从它开局', async () => {
    const cache = createComboCache()
    const { transport } = onlyWorking('lite=js@https://bbs.nga.cn')
    const request = { path: 'thread.php', query: { fid: 650 } }
    await rotatingFetcher(transport, cache)(request)
    expect(cache.get(interfaceKeyOf(request))).toBeDefined()

    const blocked = fakeTransport(() => ({ status: 403, body: BLOCKED, contentType: 'text/html' }))
    await expect(rotatingFetcher(blocked.transport, cache)(request)).rejects.toMatchObject({
      kind: 'parse',
    })

    expect(cache.get(interfaceKeyOf(request))).toBeUndefined()
    expect(blocked.combos).toHaveLength(4)
  })

  it('业务错误(权限不足)不触发降级:那说明这个组合根本没被封', async () => {
    const { transport, combos } = fakeTransport(() => ({
      status: 403,
      body: utf8('{"error":{"code":8,"0":"您没有浏览该版面的权限"}}'),
    }))
    const fetchNga = rotatingFetcher(transport)

    await expect(fetchNga({ path: 'thread.php', query: { fid: 650 } })).rejects.toMatchObject({
      kind: 'server',
      message: expect.stringContaining('权限'),
    })
    expect(combos).toHaveLength(1)
  })

  it('业务错误也算这个组合是通的,一样记进缓存', async () => {
    const cache = createComboCache()
    const { transport } = fakeTransport((combo) =>
      combo === 'lite=js@https://bbs.nga.cn'
        ? { body: utf8('{"error":{"0":"2048:找不到主题"}}') }
        : { status: 403, body: BLOCKED, contentType: 'text/html' },
    )
    const request = { path: 'read.php', query: { tid: 1 } }

    await expect(rotatingFetcher(transport, cache)(request)).rejects.toMatchObject({
      kind: 'server',
    })

    expect(cache.get(interfaceKeyOf(request))).toEqual({
      format: 'jsonLite',
      host: 'https://bbs.nga.cn',
    })
  })

  it('组合数上限生效,不会让人等十几个来回', async () => {
    const { transport, combos } = fakeTransport(() => ({
      status: 403,
      body: BLOCKED,
      contentType: 'text/html',
    }))
    const fetchNga = createNgaFetcher({
      transport,
      host: HOSTS[0],
      strategies: [createFormatRotationStrategy({ maxAttempts: 3 })],
    })

    await expect(fetchNga({ path: 'thread.php' })).rejects.toMatchObject({ kind: 'parse' })
    expect(combos).toHaveLength(3)
  })

  it('每次重试前重建 HTTP client(第一次用现成的)', async () => {
    let built = 0
    const { transport } = onlyWorking('lite=js@https://ngabbs.com')
    const fetchNga = createNgaFetcher({
      createTransport: () => {
        built += 1
        return transport
      },
      host: HOSTS[0],
      strategies: [
        createFormatRotationStrategy({ formats: FORMATS, hosts: HOSTS, maxAttempts: 10 }),
      ],
    })

    await fetchNga({ path: 'thread.php' })

    // 建 fetcher 时一次 + 第 2/3/4 次尝试各重建一次
    expect(built).toBe(4)
  })

  it('调用方指定的域名排第一,被封了照样往下轮换', async () => {
    const { transport, combos } = onlyWorking('__output=8@https://bbs.nga.cn')
    const fetchNga = rotatingFetcher(transport)

    await fetchNga({ path: 'thread.php', host: 'https://nga.178.com' })

    expect(combos[0]).toBe('__output=8@https://nga.178.com')
    expect(combos.at(-1)).toBe('__output=8@https://bbs.nga.cn')
  })

  it('取消请求不当被封,一次就停', async () => {
    const controller = new AbortController()
    const combos: string[] = []
    const transport: HttpTransport = (request) => {
      combos.push(comboOf(request))
      controller.abort()
      return Promise.reject(new DOMException('Aborted', 'AbortError'))
    }
    const fetchNga = rotatingFetcher(transport)

    await expect(
      fetchNga({ path: 'thread.php', signal: controller.signal }),
    ).rejects.toMatchObject({ kind: 'network', retryable: false })
    expect(combos).toHaveLength(1)
  })
})

/**
 * 「未登录」这一档单独测：它长得像服务端语义错误，实际是这一发请求没带上身份。
 * 真机取证（2026-08-13，小米 25113PN0EC）：冷启动后第一个版块约 1/6 概率直接
 * 报「1:未登录」，手点重试（= 忘掉组合重来）立刻就好——因为 okhttp 的 cookie jar
 * 按域名存，换个域名我们自己拼的 Cookie 头就不会被顶掉（见 auth.ts 文件头）。
 */
describe('createFormatRotationStrategy · 服务端说「未登录」', () => {
  const UNAUTHED = utf8('{"error":{"code":1,"0":"未登录"},"time":1}')
  const CREDENTIALS = { uid: '10000001', token: 'cid-a' }

  /** 只有 `only` 这个组合认得出身份，其余一律回「未登录」。 */
  function onlyAuthed(only: string) {
    return fakeTransport((combo) => (combo === only ? { body: OK } : { body: UNAUTHED }))
  }

  it('手上有凭证时继续换组合,直到某个域名认出身份', async () => {
    const { transport, combos } = onlyAuthed('lite=js@https://ngabbs.com')
    const fetchNga = createNgaFetcher({
      transport,
      host: HOSTS[0],
      getCredentials: () => CREDENTIALS,
      strategies: [createFormatRotationStrategy({ formats: FORMATS, hosts: HOSTS, maxAttempts: 10 })],
    })

    const result = await fetchNga({ path: 'thread.php', query: { fid: 650 } })

    expect(result.via).toBe('format-rotation')
    expect(combos).toEqual([
      '__output=8@https://bbs.nga.cn',
      'lite=js@https://bbs.nga.cn',
      '__output=8@https://ngabbs.com',
      'lite=js@https://ngabbs.com',
    ])
  })

  it('丢身份的那个组合不会被记进缓存,下一次不从它开局', async () => {
    const cache = createComboCache()
    const { transport, combos } = onlyAuthed('lite=js@https://ngabbs.com')
    const fetchNga = createNgaFetcher({
      transport,
      comboCache: cache,
      host: HOSTS[0],
      getCredentials: () => CREDENTIALS,
      strategies: [createFormatRotationStrategy({ formats: FORMATS, hosts: HOSTS, maxAttempts: 10 })],
    })
    const request = { path: 'thread.php', query: { fid: 650 } }

    await fetchNga(request)
    expect(cache.get(interfaceKeyOf(request))).toEqual({
      format: 'jsonLite',
      host: 'https://ngabbs.com',
    })

    combos.length = 0
    await fetchNga(request)
    expect(combos).toEqual(['lite=js@https://ngabbs.com'])
  })

  it('游客态不换组合:哪个域名都没 cookie,白跑一整轮只会把错误页拖慢', async () => {
    const { transport, combos } = onlyAuthed('lite=js@https://ngabbs.com')
    const fetchNga = createNgaFetcher({
      transport,
      host: HOSTS[0],
      getCredentials: () => null,
      strategies: [createFormatRotationStrategy({ formats: FORMATS, hosts: HOSTS, maxAttempts: 10 })],
    })

    await expect(fetchNga({ path: 'thread.php', query: { fid: 650 } })).rejects.toMatchObject({
      kind: 'server',
      message: '未登录',
    })
    expect(combos).toEqual(['__output=8@https://bbs.nga.cn'])
  })

  /**
   * 但「不换组合」不等于「掐死整条链」。真机取证（2026-08-13）：游客态打开帖子，
   * 直连报未登录，而点「用网页版打开」正文完整渲染——网页兜底本来就能拿到这一页，
   * 只是以前错误被判成不可重试，`runStrategyChain` 在轮到它之前就抛了。
   */
  it('游客态的未登录仍然可重试,好让链上后面的网页兜底接手', async () => {
    const { transport, combos } = onlyAuthed('(没有能用的组合)')
    const rescue = {
      name: 'web-fallback',
      run: () =>
        Promise.resolve({
          ok: true as const,
          result: { root: { data: { rescued: true } }, data: { rescued: true }, via: 'web-fallback' },
        }),
    }
    const fetchNga = createNgaFetcher({
      transport,
      host: HOSTS[0],
      getCredentials: () => null,
      strategies: [
        createFormatRotationStrategy({ formats: FORMATS, hosts: HOSTS, maxAttempts: 10 }),
        rescue,
      ],
    })

    const result = await fetchNga({ path: 'read.php', query: { tid: 1 } })

    expect(result.via).toBe('web-fallback')
    // 只发了一次直连就让位给兜底,没有白跑一整轮组合
    expect(combos).toEqual(['__output=8@https://bbs.nga.cn'])
  })

  it('所有组合都说未登录时,报给用户的仍是服务端原话', async () => {
    const { transport, combos } = onlyAuthed('(没有能用的组合)')
    const fetchNga = createNgaFetcher({
      transport,
      host: HOSTS[0],
      getCredentials: () => CREDENTIALS,
      strategies: [createFormatRotationStrategy({ formats: FORMATS, hosts: HOSTS, maxAttempts: 10 })],
    })

    await expect(fetchNga({ path: 'thread.php', query: { fid: 650 } })).rejects.toMatchObject({
      kind: 'server',
      message: '未登录',
    })
    expect(combos).toHaveLength(4)
  })
})

File: app/src/test/kotlin/com/chasel/ng2n/core/net/FormatRotationTest.kt
package com.chasel.ng2n.core.net

import com.chasel.ng2n.core.net.strategies.FormatRotationStrategy
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * 逐条移植自 `src/core/net/strategies/format-rotation.test.ts`
 * (`格式参数 × 域名的组合枚举` 11 条 + `服务端说「未登录」` 5 条 = 16 条,全部移植)。
 */
class FormatRotationTest {

  private val hosts = listOf("https://bbs.nga.cn", "https://ngabbs.com")
  private val formats = listOf(ResponseFormat.JSON, ResponseFormat.JSON_LITE)

  private fun rotating(
    transport: Transport,
    cache: ComboCache = InMemoryComboCache(),
    credentials: FakeCredentials = FakeCredentials(),
    maxAttempts: Int = 10,
  ) = testClient(
    transport,
    strategies = listOf(FormatRotationStrategy(formats, hosts, maxAttempts)),
    comboCache = cache,
    credentials = credentials,
    settings = FakeSettings(hostValue = hosts[0]),
  )

  @Test
  fun `封禁响应触发按序降级,直到某个组合通了`() = runTest {
    val transport = onlyWorking("lite=js@https://ngabbs.com")

    val result = rotating(transport).execute(readRequest("thread.php", queryOf("fid" to 650)))

    assertEquals("format-rotation", result.via)
    assertEquals(
      listOf(
        "__output=8@https://bbs.nga.cn",
        "lite=js@https://bbs.nga.cn",
        "__output=8@https://ngabbs.com",
        "lite=js@https://ngabbs.com",
      ),
      transport.combos(),
    )
  }

  @Test
  fun `成功的组合按接口 key 记进缓存,下一次直接命中不再从头试`() = runTest {
    val cache = InMemoryComboCache()
    val transport = onlyWorking("lite=js@https://ngabbs.com")
    val client = rotating(transport, cache)
    val request = readRequest("thread.php", queryOf("fid" to 650))

    client.execute(request)
    assertEquals(
      FetchCombo(ResponseFormat.JSON_LITE, "https://ngabbs.com"),
      cache.get(interfaceKeyOf(request)),
    )

    transport.clear()
    client.execute(request)
    assertEquals(listOf("lite=js@https://ngabbs.com"), transport.combos())
  }

  @Test
  fun `缓存是按接口分的·另一个接口还得自己从头试`() = runTest {
    val cache = InMemoryComboCache()
    val transport = onlyWorking("lite=js@https://bbs.nga.cn")
    val client = rotating(transport, cache)

    client.execute(readRequest("nuke.php", queryOf("__lib" to "noti", "__act" to "get_all")))
    transport.clear()
    client.execute(readRequest("nuke.php", queryOf("__lib" to "ucp", "__act" to "get")))

    assertEquals(
      listOf("__output=8@https://bbs.nga.cn", "lite=js@https://bbs.nga.cn"),
      transport.combos(),
    )
  }

  @Test
  fun `全组合都被封时,缓存里那个也清掉,免得下次还从它开局`() = runTest {
    val cache = InMemoryComboCache()
    val request = readRequest("thread.php", queryOf("fid" to 650))
    rotating(onlyWorking("lite=js@https://bbs.nga.cn"), cache).execute(request)
    assertNotNull(cache.get(interfaceKeyOf(request)))

    val allBlocked = RecordingTransport { blocked() }
    val error = assertThrowsNga { rotating(allBlocked, cache).execute(request) }

    assertEquals(NgaErrorKind.PARSE, error.kind)
    assertNull(cache.get(interfaceKeyOf(request)))
    assertEquals(4, allBlocked.requests.size)
  }

  @Test
  fun `业务错误(权限不足)不触发降级·那说明这个组合根本没被封`() = runTest {
    val transport = RecordingTransport {
      FakeResponse(
        status = 403,
        body = utf8("""{"error":{"code":8,"0":"您没有浏览该版面的权限"}}"""),
      )
    }

    val error = assertThrowsNga {
      rotating(transport).execute(readRequest("thread.php", queryOf("fid" to 650)))
    }

    assertEquals(NgaErrorKind.SERVER, error.kind)
    assertTrue(error.text.contains("权限"), error.text)
    assertEquals(1, transport.requests.size)
  }

  @Test
  fun `业务错误也算这个组合是通的,一样记进缓存`() = runTest {
    val cache = InMemoryComboCache()
    val transport = RecordingTransport {
      if (comboOf(it) == "lite=js@https://bbs.nga.cn") {
        ok("""{"error":{"0":"2048:找不到主题"}}""")
      } else {
        blocked()
      }
    }
    val request = readRequest("read.php", queryOf("tid" to 1))

    val error = assertThrowsNga { rotating(transport, cache).execute(request) }

    assertEquals(NgaErrorKind.SERVER, error.kind)
    assertEquals(
      FetchCombo(ResponseFormat.JSON_LITE, "https://bbs.nga.cn"),
      cache.get(interfaceKeyOf(request)),
    )
  }

  @Test
  fun `组合数上限生效,不会让人等十几个来回`() = runTest {
    val transport = RecordingTransport { blocked() }
    val client = testClient(
      transport,
      strategies = listOf(FormatRotationStrategy(maxAttempts = 3)),
      settings = FakeSettings(hostValue = hosts[0]),
    )

    val error = assertThrowsNga { client.execute(readRequest("thread.php")) }
    assertEquals(NgaErrorKind.PARSE, error.kind)
    assertEquals(3, transport.requests.size)
  }

  @Test
  fun `每次重试前重建 HTTP client(第一次用现成的)`() = runTest {
    val transport = onlyWorking("lite=js@https://ngabbs.com")
    val factory = CountingTransportFactory(transport)
    val client = testClient(
      transport,
      transports = factory,
      strategies = listOf(FormatRotationStrategy(formats, hosts, maxAttempts = 10)),
      settings = FakeSettings(hostValue = hosts[0]),
    )

    client.execute(readRequest("thread.php"))

    // 建 context 时一次 + 第 2/3/4 次尝试各重建一次
    assertEquals(4, factory.built)
  }

  @Test
  fun `调用方指定的域名排第一,被封了照样往下轮换`() = runTest {
    val transport = onlyWorking("__output=8@https://bbs.nga.cn")

    rotating(transport).execute(readRequest("thread.php", host = "https://nga.178.com"))

    assertEquals("__output=8@https://nga.178.com", transport.combos().first())
    assertEquals("__output=8@https://bbs.nga.cn", transport.combos().last())
  }

  @Test
  fun `取消请求不当被封,一次就停`() = runTest {
    // TS 版是「network / retryable=false」;Kotlin 走协程取消,原样抛
    // (见 NgaClientResponseTest 同名用例的注释)。「一次就停」这一半不变。
    var attempts = 0
    val transport = Transport {
      attempts += 1
      throw kotlinx.coroutines.CancellationException("Aborted")
    }
    var cancelled = false
    try {
      rotating(transport).execute(readRequest("thread.php"))
    } catch (_: kotlinx.coroutines.CancellationException) {
      cancelled = true
    }

    assertTrue(cancelled)
    assertEquals(1, attempts)
  }

  @Test
  fun `缓存里那个组合当场失手就先摘掉(自愈)`() = runTest {
    // ADR-0002 第 2 条:组合半通不通时缓存要有出口,不能只靠「全组合都失败」
    val cache = InMemoryComboCache()
    cache.remember("thread.php", FetchCombo(ResponseFormat.JSON_LITE, "https://ngabbs.com"))
    val transport = onlyWorking("__output=8@https://bbs.nga.cn")

    rotating(transport, cache).execute(readRequest("thread.php"))

    assertEquals(
      FetchCombo(ResponseFormat.JSON, "https://bbs.nga.cn"),
      cache.get("thread.php"),
    )
    assertEquals("lite=js@https://ngabbs.com", transport.combos().first())
  }

  // ────────────────────────────────────────────────────────────────────────────
  // 服务端说「未登录」:它长得像服务端语义错误,实际是这一发请求没带上身份。
  // 真机取证(2026-08-13,小米 25113PN0EC):冷启动后第一个版块约 1/6 概率直接报
  // 「1:未登录」,手点重试(= 忘掉组合重来)立刻就好 —— 因为 cookie jar 按域名存,
  // 换个域名我们自己拼的 Cookie 头就不会被顶掉。
  // ────────────────────────────────────────────────────────────────────────────

  private val unauthed = """{"error":{"code":1,"0":"未登录"},"time":1}"""

  private fun onlyAuthed(only: String) =
    RecordingTransport { if (comboOf(it) == only) ok() else ok(unauthed) }

  @Test
  fun `手上有凭证时继续换组合,直到某个域名认出身份`() = runTest {
    val transport = onlyAuthed("lite=js@https://ngabbs.com")
    val client = rotating(transport, credentials = FakeCredentials(signedIn = ALICE))

    val result = client.execute(readRequest("thread.php", queryOf("fid" to 650)))

    assertEquals("format-rotation", result.via)
    assertEquals(
      listOf(
        "__output=8@https://bbs.nga.cn",
        "lite=js@https://bbs.nga.cn",
        "__output=8@https://ngabbs.com",
        "lite=js@https://ngabbs.com",
      ),
      transport.combos(),
    )
  }

  @Test
  fun `丢身份的那个组合不会被记进缓存,下一次不从它开局`() = runTest {
    val cache = InMemoryComboCache()
    val transport = onlyAuthed("lite=js@https://ngabbs.com")
    val client = rotating(transport, cache, FakeCredentials(signedIn = ALICE))
    val request = readRequest("thread.php", queryOf("fid" to 650))

    client.execute(request)
    assertEquals(
      FetchCombo(ResponseFormat.JSON_LITE, "https://ngabbs.com"),
      cache.get(interfaceKeyOf(request)),
    )

    transport.clear()
    client.execute(request)
    assertEquals(listOf("lite=js@https://ngabbs.com"), transport.combos())
  }

  @Test
  fun `游客态不换组合·哪个域名都没 cookie,白跑一整轮只会把错误页拖慢`() = runTest {
    val transport = onlyAuthed("lite=js@https://ngabbs.com")

    val error = assertThrowsNga {
      rotating(transport).execute(readRequest("thread.php", queryOf("fid" to 650)))
    }

    assertEquals(NgaErrorKind.SERVER, error.kind)
    assertEquals("未登录", error.text)
    assertEquals(listOf("__output=8@https://bbs.nga.cn"), transport.combos())
  }

  /**
   * 但「不换组合」不等于「掐死整条链」。真机取证(2026-08-13):游客态打开帖子,
   * 直连报未登录,而点「用网页版打开」正文完整渲染 —— 网页兜底本来就能拿到这一页,
   * 只是以前错误被判成不可重试,`runStrategyChain` 在轮到它之前就抛了。
   */
  @Test
  fun `游客态的未登录仍然可重试,好让链上后面的网页兜底接手`() = runTest {
    val transport = onlyAuthed("(没有能用的组合)")
    val rescue = StubStrategy("web-fallback")
    val client = testClient(
      transport,
      strategies = listOf(FormatRotationStrategy(formats, hosts, maxAttempts = 10), rescue),
      settings = FakeSettings(hostValue = hosts[0]),
    )

    val result = client.execute(readRequest("read.php", queryOf("tid" to 1)))

    assertEquals("web-fallback", result.via)
    // 只发了一次直连就让位给兜底,没有白跑一整轮组合
    assertEquals(listOf("__output=8@https://bbs.nga.cn"), transport.combos())
  }

  @Test
  fun `所有组合都说未登录时,报给用户的仍是服务端原话`() = runTest {
    val transport = onlyAuthed("(没有能用的组合)")

    val error = assertThrowsNga {
      rotating(transport, credentials = FakeCredentials(signedIn = ALICE))
        .execute(readRequest("thread.php", queryOf("fid" to 650)))
    }

    assertEquals(NgaErrorKind.SERVER, error.kind)
    assertEquals("未登录", error.text)
    assertEquals(4, transport.requests.size)
  }
}
